hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/CassAlertsServiceImpl.java

```diff
@@ -0,0 +1,849 @@
+
+            initPreparedStatements();
+
+            GsonBuilder gsonBuilder = new GsonBuilder();
+            gsonBuilder.registerTypeHierarchyAdapter(ConditionEval.class, new GsonAdapter<ConditionEval>());
+            gson = gsonBuilder.create();
+
+            reload();
+        } catch (Throwable t) {
+            msgLog.errorCannotInitializeAlertsService(t.getMessage());
+        }
+    }
+
+    private void initPreparedStatements() throws Exception {
+        if (insertAlert == null) {
+            insertAlert = session.prepare("INSERT INTO " + keyspace + ".alerts " +
+                    "(alertId, payload) VALUES (?, ?) ");
+        }
+        if (insertAlertTrigger == null) {
+            insertAlertTrigger = session.prepare("INSERT INTO " + keyspace + ".alerts_triggers " +
+                    "(alertId, triggerId) VALUES (?, ?) ");
+        }
+        if (insertAlertCtime == null) {
+            insertAlertCtime = session.prepare("INSERT INTO " + keyspace + ".alerts_ctimes " +
+                    "(alertId, ctime) VALUES (?, ?) ");
+
+        }
+        if (insertAlertStatus == null) {
+            insertAlertStatus = session.prepare("INSERT INTO " + keyspace + ".alerts_statuses " +
+                    "(alertId, status) VALUES (?, ?) ");
+        }
+        if (updateAlert == null) {
+            updateAlert = session.prepare("UPDATE " + keyspace + ".alerts " +
+                    "SET payload = ? WHERE alertId = ? ");
+        }
+        if (selectAlertStatus == null) {
+            selectAlertStatus = session.prepare("SELECT alertId, status FROM " + keyspace + ".alerts_statuses " +
+                    "WHERE alertId = ? ALLOW FILTERING ");
+        }
+        if (deleteAlertStatus == null) {
+            deleteAlertStatus = session.prepare("DELETE FROM " + keyspace + ".alerts_statuses " +
+                    "WHERE alertId = ? AND status = ?");
+        }
+    }
+
+    public void addAlerts(Collection<Alert> alerts) throws Exception {
+        if (alerts == null) {
+            throw new IllegalArgumentException("Alerts must be not null");
+        }
+        if (session == null) {
+            throw new RuntimeException("Cassandra session is null");
+        }
+        if (insertAlert == null) {
+            throw new RuntimeException("insertAlert PreparedStatement is null");
+        }
+        try {
+            for (Alert a : alerts) {
+                session.execute(insertAlert.bind(a.getAlertId(), toJson(a)));
+                session.execute(insertAlertTrigger.bind(a.getAlertId(), a.getTriggerId()));
+                session.execute(insertAlertCtime.bind(a.getAlertId(), a.getCtime()));
+                session.execute(insertAlertStatus.bind(a.getAlertId(), a.getStatus().name()));
+            }
+        } catch (Exception e) {
+            msgLog.errorDatabaseException(e.getMessage());
+            throw e;
+        }
+    }
+
+    public List<Alert> getAlerts(AlertsCriteria criteria) throws Exception {
```

Comment:
This is pretty big method with a lot going on. I may have missed some things, but it basically looks like it is doing,

* Fetch alert ids from alert_triggers table if filtering by trigger ids
* Fetch alert ids from alerts_ctimes table if filtering by ctime
* Fetch alert ids from alerts_statuses table if filtering by status
* Merge all alert ids
* Fetch alerts from alerts table

Let me first share a few, general observations. With data modeling/schema design, you want to design in such a way that optimizes for your reads. In the ideal scenario when fetching a single entity or even multiple entities,  you can satisfy the request by reading a single partition. Accomplishing this usually requires denormalizing your data model and performing additional filtering/merging on the client side. Keep in mind, I am not suggesting that doing this is easy :)

Here is an idea for an alternative approach. In each of the filter tables, alerts_triggers, alerts_ctimes, alerts_statuses, store the payload in addition to the alert id. Then for each filter that is used, you can query against each of the filter tables in parallel, merging the results on the client side using AND or OR semantics as appropriate or as dictated by the AlertCriteria. 

If an alert payload is large and/or there is unbounded growth of them, then fewer filter tables with more client-side filtering might be better. For example, maybe do something like,

```
CREATE TABLE alerts_triggers (
    triggerId text,
    ctime bigint,
    alertId text,
    status text,
    PRIMARY KEY (triggerId, ctime, alertId)
);
```
Then you can do,

```
SELECT alertId 
FROM alerts_triggers 
WHERE triggerId = ? AND ctime >= ? AND ctime <= ?
```

You would want to execute these in parallel, and then you could filter by status on the client side. Lastly, you should be using prepared statements.